### PR TITLE
energy fixer impl for comments

### PR DESCRIPTION
**Pull request by oksanaguba**

### CODE REVIEW

**Comment by oksanaguba** [comment]
the last commit is bfb on chryalis, no threads, 64,32, 17, and 3 ranks
```
[ac.onguba@chrlogin2 apr21]$ zgrep hash ~/ess/apr21b/run/e3sm.log.* | grep "738 "
/home/ac.onguba/ess/apr21b/run/e3sm.log.731963.250421-153625.gz: 0: bfbhash>            738 b1fcc15e61c1c6d8 (Hommexx)
/home/ac.onguba/ess/apr21b/run/e3sm.log.731969.250421-160131.gz: 0: bfbhash>            738 b1fcc15e61c1c6d8 (Hommexx)
/home/ac.onguba/ess/apr21b/run/e3sm.log.732649.250422-131834.gz: 0: bfbhash>            738 b1fcc15e61c1c6d8 (Hommexx)
/home/ac.onguba/ess/apr21b/run/e3sm.log.732655.250422-140013.gz: 0: bfbhash>            738 b1fcc15e61c1c6d8 (Hommexx)
```

**Comment by oksanaguba** [comment]
here is example of log entries for the fixer 
```
EAMxx:: energy fixer, PB T tendency -0.000845
EAMxx:: energy fixer, total energy before fix 32484931544.607445
EAMxx:: energy fixer, rel energy error after fix 4.32541348833857e-17
```
PB T tendency is the peanut-buttered value of T tend applied everywhere in the domain.

**Comment by oksanaguba** [comment]
@bartgol @ambrad  -- here is a draft for the fixer (but without 4-columns summation). do you have any suggestions so far?

**Comment by oksanaguba** [comment]
Here are WIP items so far:

-- Put line 'atm_proc->add_column_conservation_check(conservation_check, fail_handling_type);' in the same or similar if-statement as for column checks (i haven't figure out whether the same if-statements are enough).

-- Decide what to do about 4-columns summation

-- Added 2 new views, could i get away without them?
  m_new_energy_for_fixer = view_1d<Real> ("new energy fixer", m_num_cols);
  m_energy_change = view_1d<Real> ("energy change fixer", m_num_cols);

-- Field Layouts/views code is rather cumbersome, but maybe that is common in AD.

-- I haven't tried standalone scream, there is a question what to do about components/eamxx/src/share/util/eamxx_repro_sum_mod.F90 and '#ifdef CAM'.

**Review by mahf708 — COMMENTED** [review]
Looks fine to me. I made one comment about potentially eaiser code, but not important, so don't worry about it (Luca or I can help with it later).

The overall logic looks solid to me. Are you happy with the results you're seeing? Are things "correct" to the degree you're happy?

I am not entirely clear about the 4-col summation, so I will ask privately for clarification

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 1]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+  const Real* send;
+  Real* recv;
+  Int nlocal = ncols;
+  Int ncount = 1;
+  auto s1_host = field_version_s1.get_view<const Real*,Host>();
+  auto s0_host = res.get_view<Real,Host>();
+  send = s1_host.data();
+  recv = s0_host.data();
+
+  eamxx_repro_sum(send, recv, nlocal, ncount, MPI_Comm_c2f(m_comm.mpi_comm()));
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:356]: You could do just one global reduction by combining this call to repro_sum with the next.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 2]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+
+  //reprosum vars
+  const Real* send;
+  Real* recv;
+  Int nlocal = ncols;
+  Int ncount = 1;
+  auto s1_host = field_version_s1.get_view<const Real*,Host>();
+  auto s0_host = res.get_view<Real,Host>();
+  send = s1_host.data();
+  recv = s0_host.data();
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:354]: Here I think you can use a simple `Real recv`, and then pass `&recv` to repro_sum.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 3]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+  eamxx_repro_sum(send, recv, nlocal, ncount, MPI_Comm_c2f(m_comm.mpi_comm()));
+  pb_fixer = (*recv);
+
+  if(print_debug_info) {
+    //total energy needed for relative error
+    Kokkos::parallel_for(policy, KOKKOS_LAMBDA (const KT::MemberType& team) {
+      const int i = team.league_rank();
+      field_view_s1(i) = m_current_energy(i) * area_view(i);
+    });
+    eamxx_repro_sum(send, recv, nlocal, ncount, MPI_Comm_c2f(m_comm.mpi_comm()));
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:388]: I know this is for analysis only, but you could consider combining call, also, into one overall call to repro_sum. I'm not seeing any data dependencies that would prevent that.

File: components/eamxx/src/dynamics/homme/eamxx_homme_process_interface.hpp [hunk 4]
```diff
@@ -7,6 +7,7 @@
 #include "ekat/ekat_parameter_list.hpp"
 #include "ekat/ekat_pack.hpp"
 #include "ekat/ekat_workspace.hpp"
+#include "physics/p3/p3_functions.hpp"
```
Comment (ambrad) [inline components/eamxx/src/dynamics/homme/eamxx_homme_process_interface.hpp:10]: We probably should avoid including unrelated headers.

File: components/eamxx/src/share/atm_process/atmosphere_process.cpp [hunk 5]
```diff
@@ -1121,18 +1157,44 @@ ::remove_group (const std::string& group_name, const std::string& grid_name) {
...
+  const auto& conservation_check =
+      std::dynamic_pointer_cast<MassAndEnergyColumnConservationCheck>(m_conservation.second);
+
+  //dt is needed to convert flux to change
+  conservation_check->set_dt(dt);
+  conservation_check->global_fixer(print_debug_info);
+
+  if(print_debug_info){
+    //print everything about the fixer only in debug mode
+    m_atm_logger->info("EAMxx:: energy fixer, PB T tendency " + std::to_string( conservation_check->get_pb_fixer() ) );
```
Comment (ambrad) [inline components/eamxx/src/share/atm_process/atmosphere_process.cpp:1191]: "PB T" won't mean much to others. Perhaps something like "T tendency added to each physics cell"?

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 6]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+    const auto qv_i             = ekat::subview(qv, i);
+    const auto qc_i             = ekat::subview(qc, i);
+    const auto qr_i             = ekat::subview(qr, i);
+    const auto qi_i             = ekat::subview(qi, i);
+
+    // Calculate total energy
+    m_new_energy_for_fixer(i) = compute_total_energy_on_column(team, nlevs, pseudo_density_i, T_mid_i, horiz_winds_i,
+                                                   qv_i, qc_i, qr_i, ps(i), phis(i));
+    m_energy_change(i) = compute_energy_boundary_flux_on_column(vapor_flux(i), water_flux(i), ice_flux(i), heat_flux(i))*dt;
+    field_view_s1(i) = (m_current_energy(i)-m_new_energy_for_fixer(i)-m_energy_change(i)) * area_view(i);
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:376]: This technically should use `Kokkos::single` with `PerTeam`. But it's fine for all threads to write the same value here.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 7]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+  FieldIdentifier s0_fid("s0", scalar0d, kg / kg, m_grid->name());
+  Field res(s0_fid);
+  res.allocate_view();
+  //this does not exist, why? cause it can be a LEV or a COL array?
+  //FieldLayout scalar1d = m_grid->get_1d_scalar_layout();
+  FieldLayout scalar1d{{COL}, {m_num_cols}};
+  FieldIdentifier s1_fid("s1", scalar1d, kg / kg, m_grid->name());
+
+  Field field_version_s1(s1_fid);
+  field_version_s1.allocate_view();
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:333]: I think we want to avoid allocations at each time step. There should be a memory buffer for scratch available here, but perhaps I'm wrong.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 8]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+    // Calculate total gas mass (sum dp, no water loading)
+    field_view_s1(i) = compute_gas_mass_on_column(team, nlevs, pseudo_density_i) * area_view(i);
+  });
+
+  //reprosum vars
+  const Real* send;
+  Real* recv;
+  Int nlocal = ncols;
+  Int ncount = 1;
+  auto s1_host = field_version_s1.get_view<const Real*,Host>();
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:351]: @bartgol does this do the device-to-host sync automatically when it's needed? I was expecting to see something like a `sync` call here.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 9]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+
+      // Calculate total energy
+      m_new_energy_for_fixer(i) = compute_total_energy_on_column(team, nlevs, pseudo_density_i, T_mid_i, horiz_winds_i,
+                                                   qv_i, qc_i, qr_i, ps(i), phis(i));
+      //overwrite the "new" fields with relative change
+      field_view_s1(i) = (m_current_energy(i)-m_new_energy_for_fixer(i)-m_energy_change(i));
+      field_view_s1(i) *= area_view(i);
+    });
+
+    eamxx_repro_sum(send, recv, nlocal, ncount, MPI_Comm_c2f(m_comm.mpi_comm()));
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:425]: Here it seems like there really needs to be a device-to-host sync call.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 3 above)
Comment (ambrad) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:388]: Device-to-host sync?

File: components/eamxx/cime_config/namelist_defaults_eamxx.xml [hunk 10]
```diff
@@ -187,6 +187,10 @@ be lost if SCREAM_HACK_XML is not enabled.
       <!-- Frequency in physics steps to output a global hash over the dycore's
            in-fields. <= 0 disables hashing. -->
       <BfbHash type="integer">18</BfbHash>
+      <enable_column_conservation_checks>false</enable_column_conservation_checks>
+      <enable_energy_fixer>true</enable_energy_fixer>
+      <!-- true means 2 more calls to AllReduce after dycore -->
+      <enable_energy_fixer_debug_info>true</enable_energy_fixer_debug_info>
```
Comment (ambrad) [inline components/eamxx/cime_config/namelist_defaults_eamxx.xml:193]: I think the extra debug output shouldn't default to `true`.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 11]
```diff
@@ -362,4 +528,17 @@ compute_energy_boundary_flux_on_column (const Real vapor_flux,
   return vapor_flux*(LatVap+LatIce) - (water_flux-ice_flux)*RHO_H2O*LatIce + heat_flux;
 }
 
+Real MassAndEnergyColumnConservationCheck::get_echeck(){
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:531]: You can mark this and several other methods as `const`: `:get_echeck() const {`

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 12]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+    const auto pseudo_density_i = ekat::subview(pseudo_density, i);
+    const auto T_mid_i          = ekat::subview(T_mid, i);
+    const auto horiz_winds_i    = ekat::subview(horiz_winds, i);
+    const auto qv_i             = ekat::subview(qv, i);
+    const auto qc_i             = ekat::subview(qc, i);
+    const auto qr_i             = ekat::subview(qr, i);
+    const auto qi_i             = ekat::subview(qi, i);
+
+    // Calculate total energy
+    m_new_energy_for_fixer(i) = compute_total_energy_on_column(team, nlevs, pseudo_density_i, T_mid_i, horiz_winds_i,
```
Comment (ambrad) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:373]: This value of m_new_energy_for_fixer doesn't seem to be used anywhere else. It can just be a local scalar within this kernel, it appears, and the same for the other use.

File: components/eamxx/src/dynamics/homme/eamxx_homme_process_interface.hpp (hunk 4 above)
Comment (oksanaguba) [reply components/eamxx/src/dynamics/homme/eamxx_homme_process_interface.hpp:10]: thanks, that seems to be a leftover from before

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 13]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+  field_version_s1.allocate_view();
+  auto field_view_s1 = field_version_s1.get_view<Real*>();
+
+  auto area = m_grid->get_geometry_data("area").clone();
+  auto area_view = area.get_view<const Real*>();
+
+  Kokkos::parallel_for(policy, KOKKOS_LAMBDA (const KT::MemberType& team) {
+    const int i = team.league_rank();
+    const auto pseudo_density_i = ekat::subview(pseudo_density, i);
+    // Calculate total gas mass (sum dp, no water loading)
```
Comment (bartgol) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:342]: Isn't dp wet in EAMxx?

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 14]
```diff
@@ -362,4 +528,17 @@ compute_energy_boundary_flux_on_column (const Real vapor_flux,
...
 
+Real MassAndEnergyColumnConservationCheck::get_echeck(){
+  return echeck;
+}
+
+Real MassAndEnergyColumnConservationCheck::get_total_energy_before(){
+  return total_energy_before;
+}
+
+Real MassAndEnergyColumnConservationCheck::get_pb_fixer(){
```
Comment (bartgol) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:539]: As Andrew said, we should prob find a different word than pb (I'm assuming it's "peanut butter"?). Maybe something more "mathematical", like "uniform fixer"? "mean value fixer"?

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.hpp [hunk 15]
```diff
@@ -26,10 +29,14 @@ class MassAndEnergyColumnConservationCheck: public PropertyCheck {
...
 
+  //using namespace ShortFieldTagsNames;
+  //using namespace ekat::units;
+
 public:
 
   // Constructor
-  MassAndEnergyColumnConservationCheck (const std::shared_ptr<const AbstractGrid>& grid,
+  MassAndEnergyColumnConservationCheck (const ekat::Comm& comm,
+                                        const std::shared_ptr<const AbstractGrid>& grid,
```
Comment (bartgol) [inline components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.hpp:39]: FYI, the grid already stores a comm object, so you can also get it by `grid->get_comm()`.

File: components/eamxx/src/share/util/eamxx_repro_sum_mod.F90 [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+module eamxx_repro_sum_mod
+
+  implicit none
+
+contains
+
+  subroutine eamxx_repro_sum(send, recv, nlocal, nfld, comm) bind(c)
+    use iso_c_binding, only: c_int, c_double
+
+!standalone scream -- does it have access to shared libs?
```
Comment (bartgol) [inline components/eamxx/src/share/util/eamxx_repro_sum_mod.F90:14]: Yes. But I don't think it builds `shr_reprosum_mod` in it. You should add it to [this](https://github.com/E3SM-Project/E3SM/blob/620edabb54fda2c88f0d1dbd2ee14e25ab3edaf3/components/eamxx/cmake/tpls/CsmShare.cmake#L28) list.

File: components/eamxx/src/share/util/eamxx_repro_sum_mod.F90 [hunk 17]
```diff
@@ -0,0 +1,31 @@
+#ifdef HAVE_CONFIG_H
+#include "config.h"
```
Comment (bartgol) [inline components/eamxx/src/share/util/eamxx_repro_sum_mod.F90:2]: Do we need `config.h` at all? I don't think so, no?

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 9 above)
Comment (ambrad) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:425]: Yes, exactly.

File: components/eamxx/cime_config/namelist_defaults_eamxx.xml (hunk 10 above)
Comment (oksanaguba) [reply components/eamxx/cime_config/namelist_defaults_eamxx.xml:193]: thanks, i do not want functionality of ${DEBUG} there, switching to false.

File: components/eamxx/src/share/atm_process/atmosphere_process.cpp (hunk 5 above)
Comment (oksanaguba) [reply components/eamxx/src/share/atm_process/atmosphere_process.cpp]: done thanks

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 7 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: done thanks

File: components/eamxx/src/dynamics/homme/eamxx_homme_process_interface.cpp [hunk 18]
```diff
@@ -749,11 +761,28 @@ void HommeDynamics::homme_post_process (const double dt) {
...
-  });
+  }); //op()
 
   // Apply Rayleigh friction to update temperature and horiz_winds
   rayleigh_friction_apply(dt);
-}
+
+  if (has_energy_fixer()) {
+
+    const auto& vapor_flux = get_field_out("vapor_flux").get_view<Real*>();
```
[the remark this answers is not in the document]
Comment (oksanaguba) [reply components/eamxx/src/dynamics/homme/eamxx_homme_process_interface.cpp]: done thanks

File: components/eamxx/src/share/CMakeLists.txt [hunk 19]
```diff
@@ -41,6 +41,7 @@ set(SHARE_SRC
   util/eamxx_timing.cpp
   util/eamxx_utils.cpp
   util/eamxx_bfbhash.cpp
+  util/eamxx_repro_sum_mod.F90
```
[the remark this answers is not in the document]
Comment (oksanaguba) [reply components/eamxx/src/share/CMakeLists.txt:44]: done thanks

File: components/eamxx/src/share/atm_process/atmosphere_process_group.cpp [hunk 20]
```diff
@@ -319,7 +318,17 @@ setup_column_conservation_checks (const std::shared_ptr<MassAndEnergyColumnConse
 
     // If all conditions are satisfied, add as postcondition_check
     atm_proc->add_column_conservation_check(conservation_check, fail_handling_type);
-  }
+  }// for (auto atm_proc : m_atm_processes) 
+
+  for (auto atm_proc : m_atm_processes) {
+//WRONG -- this needs to be in the above block with all conditionals on fluxes, T, etc.
```
[the remark this answers is not in the document]
Comment (oksanaguba) [reply components/eamxx/src/share/atm_process/atmosphere_process_group.cpp]: done thanks

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 21]
```diff
@@ -8,7 +8,8 @@ namespace scream
 {
 
 MassAndEnergyColumnConservationCheck::
-MassAndEnergyColumnConservationCheck (const std::shared_ptr<const AbstractGrid>& grid,
+MassAndEnergyColumnConservationCheck (const ekat::Comm& comm,
```
[the remark this answers is not in the document]
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: done thanks

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp [hunk 22]
```diff
@@ -278,6 +283,152 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
 }
 
 
+extern "C"
+void eamxx_repro_sum(const Real* send, Real* recv,
```
[the remark this answers is not in the document]
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: done thanks

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 13 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp:337]: Yes dp includes dry air and vapor, but no water. To distribute excess energy, i need a total integral of mass variable (dp), but it wasn't available before, as total mass computed elsewhere includes water.

File: components/eamxx/src/share/property_checks/mass_and_energy_conservation_check.cpp [hunk 23]
```diff
@@ -277,9 +294,159 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+  field_version_s1.sync_to_host(); 
+  eamxx_repro_sum(send, &recv, nlocal, ncount, MPI_Comm_c2f(m_comm.mpi_comm()));
+  field_version_s1.sync_to_dev();
+  pb_fixer = recv;
+
+  if(print_debug_info) {
+    //total energy needed for relative error
+    Kokkos::parallel_for(policy, KOKKOS_LAMBDA (const KT::MemberType& team) {
+      const int i = team.league_rank();
+      Kokkos::single(Kokkos::PerThread(team),[&]() {
```
Comment (bartgol) [inline components/eamxx/src/share/property_checks/mass_and_energy_conservation_check.cpp:389]: You prob want `Kokkos::PerTeam` here.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 1 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: i will do it in the next PR.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 2 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: thanks, done

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 3 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: done for sync, combining calls is for the next PR.

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 6 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: thanks, done

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 11 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: thanks, done

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp (hunk 12 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.cpp]: thanks, done

File: components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.hpp (hunk 15 above)
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_column_conservation_check.hpp:37]: thanks, will do it in the next PR.

File: components/eamxx/src/share/util/eamxx_repro_sum_mod.F90 (hunk 17 above)
Comment (oksanaguba) [reply components/eamxx/src/share/util/eamxx_repro_sum_mod.F90]: removed, thanks.

File: components/eamxx/src/share/property_checks/mass_and_energy_conservation_check.cpp [hunk 24]
```diff
@@ -277,9 +294,159 @@ PropertyCheck::ResultAndMsg MassAndEnergyColumnConservationCheck::check() const
...
+      const auto pseudo_density_i = ekat::subview(pseudo_density, i);
+      const auto T_mid_i          = ekat::subview(T_mid, i);
+      const auto horiz_winds_i    = ekat::subview(horiz_winds, i);
+      const auto qv_i             = ekat::subview(qv, i);
+      const auto qc_i             = ekat::subview(qc, i);
+      const auto qr_i             = ekat::subview(qr, i);
+      const auto qi_i             = ekat::subview(qi, i);
+
+      // Calculate total energy
+      const auto m_new_energy_for_fixer = compute_total_energy_on_column(team, nlevs, pseudo_density_i, T_mid_i, horiz_winds_i,
```
[the remark this answers is not in the document]
Comment (oksanaguba) [reply components/eamxx/src/share/property_checks/mass_and_energy_conservation_check.cpp]: thanks, those are static (Conrad's original code) in *hpp file.

**Comment by oksanaguba** [comment]
Going back to this -- CI output looks ok (I will look at formatting suggestions separately). I'd like to keep 3 reviewers' suggestions for the next PR -- combining calls in reprosum, using grid->comm, and renaming PB with something meaningful.

Also, once this is approved, I may need to do a new branch with cleaner history. Do you want to look at this again, @bartgol @ambrad @mahf708 ? I took Luca's advice and used CI testing, and did not run any testing by myself.

**Comment by ambrad** [comment]
Does this feature have an associated test? If not, I recommend adding one.

It's up to you whether to run tests on machines. I personally like to run a good CIME test across machines so there aren't GPU-machine surprises when I integrate one of my PRs.

Re: git history, note you can save this current branch to a new branch in your fork, then rebase this one and force-push it. That will let you keep the same PR and also save this version of the branch.

**Comment by bartgol** [comment]
Re: running tests. As Andrew said, it's up to you. Using CI saves me initial debugging, but if you want to be extra safe, running a CIME case on pm-gpu can be beneficial. Given the nature of your changes, I don't expect any surprise compared to the CI tests, provided that you enabled energy fixer in one of the standalone tests (so that CI _did_ test the fixer on GPU as part of the eamxx-sa workflow).

Re: history. Unless you do want to keep a copy of this branch as is (in which case, follow Andrew advice and create a bkp branche before doing an interactive rebase), I would simply do `git rebase -i origin/master` and squash/reword/reorg commits as you see fit. Imho, given the size of the PR, something in the range of 4-7 commits seem reasonable.

Edit: I just noticed that there is no standalone test that triggers the fixer. Perhaps it could be a good idea to enable it in one of the tests in `tests/multi-process/dynamics_physics/*` folders? E.g., `homme_shoc_cld_p3_rrtmgp/inputs.yaml` already enables col conservation checks, so maybe we can also enable the fixer there? Or does the fixer _require_ surface coupling to be present/active? If that's the case, then maybe running a GPU test before integrating may help reveal any GPU error, since no CI test runs CIME cases on GPU (for now).

**Comment by oksanaguba** [comment]
thanks, i will try to add a fixer to a standalone test (right now it is not working because probably fluxes are un-inited instead of being zero), but i do not think this is the type of test Andrew meant, because it does not test the fixer itself. I can potentially turn on debug diagn for the fixer in one of CIME tests, and then use python to parse that output like it was done for some SL/energy tests before. @ambrad ?

**Comment by ambrad** [comment]
I just meant that I recommend that this PR include a test that can run in the nightlies. What form it takes is up to you.

**Comment by oksanaguba** [comment]
For a standalone test, say, `homme_shoc_cld_p3_rrtmgp_np1` , dp is not init-ed properly before homme when compute_current_energy is called. we'll try to fix that.

**Comment by mahf708** [comment]
> For a standalone test, say, `homme_shoc_cld_p3_rrtmgp_np1` , dp is not init-ed properly before homme when compute_current_energy is called. we'll try to fix that.

You can also add a new test to these, call it `homme_shoc_cld_p3_rrtmgp_w_fixer` or so. In fact, that's what I would recommend doing first. If things are stable and good after a few months, we can remove `homme_shoc_cld_p3_rrtmgp` and keep `homme_shoc_cld_p3_rrtmgp_w_fixer` as superseding of the two. 

I will take a closer look at the entirety of the PR soon (tmrw or day after) with fresh eyes

**Comment by oksanaguba** [comment]
dp not initialized would probably be an issue in tests that run not [homme, physics], but [physics, homme], are there such tests? what do "pure physics" tests do?

**Review by bartgol — APPROVED** [review]
Looks good. Maybe I would change the PB name to something that ppl understand, but it's not that important (especially since you are planning to work more on this later, IIUC).

**Comment by oksanaguba** [comment]
To rephrase Luca's fix for un-inited dp (for my own benefit) -- since dp on phys grid did not have a time stamp, IO was filling it with FillValue before the 1st call to energy (IO does 1st instant output at init). This behavior will change soon, IO won't be overwriting fields.

**Comment by oksanaguba** [comment]
PM ERS run passed restart stage: 
```
/pscratch/sd/o/onguba/e3sm_scratch/pm-gpu/ERS_Lh4.ne4_ne4.F2010-SCREAMv1.pm-gpu_gnugpu.eamxx-output-preset-1--eamxx-fixer_debug_output.20250714_121120_6i3eza/run
```

**Comment by bartgol** [comment]
> To rephrase Luca's fix for un-inited dp (for my own benefit) -- since dp on phys grid did not have a time stamp, IO was filling it with FillValue before the 1st call to energy (IO does 1st instant output at init). This behavior will change soon, IO won't be overwriting fields.

PR #7397 was integrated, so the issue regarding un-inited dp should now be resolved even without commit 405e1c5

**Review by bartgol — APPROVED** [review]
